Add a tool for scraping rehearsal metrics from GCS

petr-muller opened the pull request:
Still a bit rough, but should be reviewable mostly for high-level feedback. If the overall structure is OK, I'll polish it. There's a little caching there but the efficiency is not stellar (still need to iterate over a lot of things), but I did not want to make it more complex, because the performance is not that bad (with 20 workers it takes about 40s to run).

/cc @stevekuznetsov @droslean @bbguimaraes

petr-muller committed:
Add a tool for scraping rehearsal metrics from GCS

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+		return
+	}
+	data, err := ioutil.ReadAll(reader)
+	_ = reader.Close()
```
is this a typo?

petr-muller replied:
Not really, GoLand suggested this as an explicit error ignore, so I used it. I'll probably just defer it.

stevekuznetsov replied:
Yeah i think I usually do

```go
defer func(reader io.Reader) {
    if err := reader.Close(); err != nil {
        logrus.WithError(err).Fatal("whatever")
    }
}(reader)
```

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+func (s *scraper) scrapePr(prefix string) {
+	prNumber := path.Base(prefix)
+	jobPrefix := path.Join(prefix, rehearseJobName)
+	jobPrefix = fmt.Sprintf("%s/", jobPrefix) // Needed because path.Join() removes trailing slash
```
yeah technically you should not use `path.Join` for URLs but whatever

droslean replied:
You can `url.Parse` them and use the `ResolveReference`.
```go
u, _ := url.Parse(prefix)
relative, _ := url.Parse(rehearseJobName)
jobPrefix := u.ResolveReference(relative)
```

petr-muller replied:
I think I'll go with the `path.Join` way, the object names behave more like files than URLs, so `path` operations are probably a bit more readable, too. The only problem is that the separator is a part of the object name, not an actual separator.

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+			break
+		}
+
+		// Leaf (a file), not a prefix (a dir)
```
Could be more clear like

```
// We only care about files, not dirs
```

petr-muller replied:
I guess. I wanted to avoid talking about files and dirs because they are not actually files and dirs, but meh.

stevekuznetsov replied:
I guess I meant this is a confusing comment so making it more clear with whatever appropriate words would be good

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+
+	fs.StringVar(&o.cacheDir, "cache-dir", "", "Path to a directory where scraped metrics data will be saved. If not provided, cache will not be used.")
+
+	_ = fs.Parse(os.Args[1:])
```
logrus.fatal on this one

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+	if o.cacheDir == "" {
+		o.cacheDir, err = ioutil.TempDir("", "")
+		if err != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "Failed to create a temporary directory: %v", err)
```
use `logrus.Fatal.WithError()`

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+			_, _ = fmt.Fprintf(os.Stderr, "Failed to create a temporary directory: %v", err)
+			os.Exit(1)
+		}
+		defer os.RemoveAll(o.cacheDir)
```
can this error?

petr-muller replied:
Yeah, will defer a func that will at least log the failure.

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+	progress := make(chan string, 100)
+
+	wg := sync.WaitGroup{}
+	wg.Add(20)
```
do you want this to be configurable?

petr-muller replied:
not really, but I'll extract it to a named constant at least

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+	wg := sync.WaitGroup{}
+	wg.Add(20)
+
+	for w := 0; w < 20; w++ {
```
Can you use a semaphore instead of this approach? This is the older approach before semaphores were added to the language

petr-muller replied:
Sure.

stevekuznetsov reviewed (commented):
approach LGTM

droslean commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+				scraper.scrapePr(prefix)
+				progress <- prefix
+			}
+			wg.Done()
```
this should be in `defer`

droslean commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+			prPrefixes <- attrs.Prefix
+		}
+		close(prPrefixes)
+		wg.Wait()
```
its better to close the channels out of the goroutine.

droslean commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+		go func() {
+			for prefix := range prPrefixes {
+				scraper.scrapePr(prefix)
+				progress <- prefix
```
this should be like:
```go
defer wg.Done()
for {
    select {
        prefix: = < -prPrefixes
        scraper.scrapePr(prefix)
        progress < -prefix
        case <-done:
            close(prPrefixes)
            close(progress)
            return
    }
}
}
```

droslean replied:
you need also the other channel:     done := make(chan struct{})

petr-muller replied:
Why _should_ this be like that? What's wrong? To me it looks more cluttered and needs an additional channel just to communicate the same thing that's communicated here by simply closing the channel.

droslean commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+
+			prPrefixes <- attrs.Prefix
+		}
+		close(prPrefixes)
```
Here you can send to `done` channel.

droslean commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,172 @@
+	}()
+
+	counter := 0
+	for done := range progress {
```
its a go idiom to use the `for{ select {} }` to handle also other signal channels to gracefully terminate.  
Also you can check if the channel closed. Like:
```go
 	for done, ok := range progress {
if !ok {
// we are done
os.Exit(0) 
}
}
```

petr-muller replied:
About the "check if the channel is closed" part, are you sure? The [spec](https://golang.org/ref/spec#For_range) says _If the range expression is a channel, at most one iteration variable is permitted_. It also does not make sense, I understand that when the channel is closed and all values are exhausted, the iteration simply ends...

stevekuznetsov replied:
Yeah unless you want to cancel in the middle you don't need the `range` and this will stop looping when the channel is closed

droslean reviewed (commented):
In general, this is a good start. I left some comments regarding the channels.

petr-muller committed:
Add semaphore and google storage deps for pj-rehearse-metrics

petr-muller commented:
I addressed the straightforward feedback and will polish the code a bit more.

@droslean Can you elaborate a bit further on how do you suggest to change the channel code? I'll need help understanding how the code could be improved, in particular:

Why is the `for { select {} }` pattern preferable to the simple `for value := channel` iteration? The simple iteration looks more readable to me and I haven't found any documentation that would discourage its use. You say _... idiom to use the for{ select {} } to handle also other signal channels to gracefully terminate_, but here's just one input channel, not more, which I understand is not the use case for `select`...

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,185 @@
+		go func() {
+			defer sem.Release(1)
+			for prefix := range prPrefixes {
+				scraper.scrapePr(prefix)
```
Err -- I think you need to limit concurrency _here_:

```go
sem := semaphore.NewWeighted(workersCount)

for prPrefix := range prPrefixes {
    go func(prefix string) {
		if err := sem.Acquire(ctx, 1); err != nil {
			logrus.WithError(err).Error("Failed to acquire semaphore to spawn a worker")
			prPrefixes<-prefix
			return
		}
		defer sem.Release(1)
		// work on the prefix
    }(prPrefix)
}
```

petr-muller replied:
I don't think so, acquiring in the spawned goroutine would create the (probably rare but still) opportunity to hit the final L173 `Acquire`  first, deadlocking the workers.

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,185 @@
+				continue
+			}
+
+			prPrefixes <- attrs.Prefix
```
You could also just remove the need for `prPrefixes`, run the content of this goroutine in the main thread and attempt to spawn a worker here -- we would read new prefixes as we need to instead of upfront but that's not a huge hit in cost, right?

petr-muller replied:
You mean to spawn short-lived goroutines to scrape just a single PR? It's possible, but I'm not sure it's better than the long-lived workers over the channel (not worse either - there's no per-worker expensive setup, so short-lived goroutines would be fine). 

I do not understand the comment about reading prefixes upfront, though - reading prefixes would still need to be implemented via looping over the `pullRequests` iterator, no? This goroutine reads prefixes and sends them to the channel as fast as the workers pop them, where's the upfront read?

stevekuznetsov replied:
The approach right now reads them all upfront and sends to the channel, the channel holds all of the values (sans whichever have started to be processed). Doing it just-in-time will read as we process.

stevekuznetsov commented on cmd/pj-rehearse-metrics/pj-rehearse-metrics.go:
```diff
@@ -0,0 +1,185 @@
+	}()
+
+	counter := 0
+	for done := range progress {
```
This is a bit clumsy, you could use a locked value or atomic updates instead

petr-muller replied:
You mean for the counter? Isn't it an overkill (locked value certainly is...) for what is basically a activity indicator?

petr-muller commented:
Removed WIP: I polished the code a bit, added few comments and extracted the paths-manipulating code to a separate struct, added tests for it. I haven't changed the concurrency structure, because I think it's correct and I prefer it to suggested alternatives (I answered individual suggestions inline).

stevekuznetsov reviewed (commented):
The whole point of the weighted semaphore is to not have the explicit pool -- what you're doing doesn't need the semaphore at all, you're basically just using it as a WaitGroup. The concurrency-bounded action is operating on the prefix, so we should acquire and release around that code. if every thread only ever acquires once then there's no real need for the semaphore, WaitGroup is sufficient. But, I suggested that approach in review 

/shrug
:man_facepalming:

stevekuznetsov commented:
/lgtm
